### PR TITLE
Enable nodejs dynamic provider caching by default on program side

### PR DESCRIPTION
**Pull request by EvanBoyle**
The full description and motivation of this change is outlined here: https://github.com/pulumi/pulumi/pull/6673

TLDR; this change enables caching by default for dynamic providers serialized as a part of executing a program via nodejs. This decreases memory usage and increases runtime significantly for programs making heavy use of dynamic providers. 

This is technically a breaking change as we are now caching dynamic providers based on object equality. It is technically true (but unlikely in practice) that a dynamic provider could rely on some global state, or that in-between resource creation operations that individual methods of the provider (diff/check/create/etc) could be mutated. 

Due to this, users should treat dynamic resource providers as immutable and opt to create singleton instances to get access to these performance benefits. This behavior can be disabled by setting `PULUMI_NODEJS_CACHE_DYNAMIC_PROVIDERS=false`.

### CODE REVIEW

File: sdk/nodejs/tests/automation/localWorkspace.spec.ts [hunk 1]
```diff
@@ -26,6 +25,7 @@ import {
     Stack,
     validatePulumiVersion,
 } from "../../automation";
+import { Config } from "../../index";
```
Comment (EvanBoyle) [inline sdk/nodejs/tests/automation/localWorkspace.spec.ts:28]: Unrelated change, however linting was failing both locally and in CI.

**Comment by EvanBoyle** [comment]
I'm seeing automation API tests fail due to the CLI version check in CI, however these passed locally after I built the CLI via after setting `VERSION_PREFIX=3.0.0`

**Review by stack72 — APPROVED** [review]
